R/kernel.r

```diff
@@ -381,6 +381,20 @@ main <- function(connection_file = '') {
         connection_file <- commandArgs(TRUE)[[1]]
     }
     log_debug('Starting the R kernel...')
+    # Environment which will take functions which shadow main R stuff like q()
+    # This has to be created here, because a) it didn't work when I tried to add
+    # it as a normal package item and not here in the codepath of the startup
+    # function and b) so the on.exit can detach the env when main is finished and
+    # c) it can't be created elsewhere because attach copies the env, so it would
+    # need to be reassigned afterwards...
+    .irk.env <- attach(NULL, name = "jupyter:irkernel")
+    on.exit({
+        detach("jupyter:irkernel")
+    })
+    assign(".irk.get_userenv", function() {.irk.env}, .irk.env)
```

Comment:
Naming quibble: 'user environment' is a lot like 'user namespace', which in IPython means the namespace where user code executes (i.e. typing `a = 1` defines a in the user namespace). That's not what we mean here. Can we come up with a clearer name? E.g. `get_shadowing_env`?